README.md
# PakhshKit JS Kava - Vidiun Advanced Video Analytics Plugin for the [Vidiun Player JS]

[![Build Status](https://travis-ci.org/vidiun/pakhshkit-js-kava.svg?branch=master)](https://travis-ci.org/vidiun/pakhshkit-js-kava)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

PakhshKit JS Kava plugin integrates Kava (Vidiun Advanced Video Analytics) with the [Vidiun Player JS].

The main purpose of this plugin is to track and collect various events and data about the video player.

PakhshKit JS Kava is written in [ECMAScript6], statically analysed using [Flow] and transpiled in ECMAScript5 using [Babel].

[flow]: https://flow.org/
[ecmascript6]: https://github.com/ericdouglas/ES6-Learning#articles--tutorials
[babel]: https://babeljs.io

## Getting Started

### Prerequisites

The plugin requires [PakhshKit JS Player] to be loaded first.

[vidiun player js]: https://github.com/vidiun/vidiun-player-js

### Installing

First, clone and run [yarn] to install dependencies:

[yarn]: https://yarnpkg.com/lang/en/

```
git clone https://github.com/vidiun/pakhshkit-js-kava.git
cd pakhshkit-js-kava
yarn install
```

### Building

Then, build the player

```javascript
yarn run build
```

### Embed the Library In Your Test Page

Finally, add the bundle as a script tag in your page, and initialize the player

```html
<script type="text/javascript" src="/PATH/TO/FILE/vidiun-{ovp/tv}-player.js"></script>
<script type="text/javascript" src="/PATH/TO/FILE/pakhshkit-kava.js"></script>
<div id="player-placeholder" style="height:360px; width:640px">
<script type="text/javascript">
var config = {
  targetId: 'player-placeholder',
  provider: {
    partnerId: {PARTNER_ID}
    ...
  },
  player: {
   plugins: {
     kava: {
        ...
     }
    ...
   }
   ...
  }
 ...
};
var player = VidiunPlayer.setup(config);
player.loadMedia({
  entryId: '{ENTRY_ID}'
  ...
});
</script>
```

## Documentation

### Table of Contents

- [Configuration](./docs/configuration.md)
- [Events](./docs/kava-events.md)
- [Parameters](./docs/kava-parameters.md)

## Running the Tests

Tests can be run locally via [Karma], which will run on Chrome, Firefox and Safari.

[karma]: https://karma-runner.github.io/1.0/index.html

```
yarn run test
```

You can test individual browsers:

```
yarn run test:chrome
yarn run test:firefox
yarn run test:safari
```

### And Coding Style Tests

We use ESLint [recommended set](http://eslint.org/docs/rules/) with some additions for enforcing [Flow] types and other rules.

See [ESLint config](.eslintrc.json) for full configuration.

We also use [.editorconfig](.editorconfig) to maintain consistent coding styles and settings, please make sure you comply with the styling.

## Compatibility

TBD

## Contributing

Please read [CONTRIBUTING.md](https://gist.github.com/PurpleBooth/b24679402957c63ec426) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/vidiun/pakhshkit-js-vanalytics/tags).

## License

This project is licensed under the AGPL-3.0 License - see the [LICENSE.md](LICENSE.md) file for details


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.2.5"></a>
## [0.2.5](https://github.com/vidiun/pakhshkit-js-kava/compare/v0.2.4...v0.2.5) (2018-07-02)


### Bug Fixes

* **FEC-8354:** When selecting Audio language and refresh the page, the display language returns to default (but actual audio proper) ([#18](https://github.com/vidiun/pakhshkit-js-kava/issues/18)) ([eb70a1c](https://github.com/vidiun/pakhshkit-js-kava/commit/eb70a1c))



<a name="0.2.4"></a>
## [0.2.4](https://github.com/vidiun/pakhshkit-js-kava/compare/v0.2.3...v0.2.4) (2018-06-28)


### Bug Fixes

* **FEC-8322:** In view event the position and playTimeSum are not equal when stream playing continuously ([#17](https://github.com/vidiun/pakhshkit-js-kava/issues/17)) ([1476ee2](https://github.com/vidiun/pakhshkit-js-kava/commit/1476ee2))



<a name="0.2.3"></a>
## [0.2.3](https://github.com/vidiun/pakhshkit-js-kava/compare/v0.2.2...v0.2.3) (2018-06-14)


### Bug Fixes

* **FEC-8313:** When playing audio entry playbackType is not "vod" ([#15](https://github.com/vidiun/pakhshkit-js-kava/issues/15)) ([9f70dc6](https://github.com/vidiun/pakhshkit-js-kava/commit/9f70dc6))
* **FEC-8317:** impression event and the followup events the position value is NAN ([#13](https://github.com/vidiun/pakhshkit-js-kava/issues/13)) ([8afa119](https://github.com/vidiun/pakhshkit-js-kava/commit/8afa119))
* **FEC-8320:** SOURCE_SELECTED event triggered on user selection of ABR mode ([#16](https://github.com/vidiun/pakhshkit-js-kava/issues/16)) ([3ca02c8](https://github.com/vidiun/pakhshkit-js-kava/commit/3ca02c8))



<a name="0.2.2"></a>
## [0.2.2](https://github.com/vidiun/pakhshkit-js-kava/compare/v0.2.1...v0.2.2) (2018-06-13)


### Bug Fixes

* **FEC-8305:** In Live videos there are progress events sent (11-14)  ([#9](https://github.com/vidiun/pakhshkit-js-kava/issues/9)) ([3f38242](https://github.com/vidiun/pakhshkit-js-kava/commit/3f38242))
* **FEC-8306:** No 'Play request' events are sent in web ([#10](https://github.com/vidiun/pakhshkit-js-kava/issues/10)) ([a78647f](https://github.com/vidiun/pakhshkit-js-kava/commit/a78647f))
* **FEC-8307:** Error event is triggered also when non-critical event occurs ([#8](https://github.com/vidiun/pakhshkit-js-kava/issues/8)) ([e6c0e0b](https://github.com/vidiun/pakhshkit-js-kava/commit/e6c0e0b))
* **FEC-8308:** audio track changes are not reproted ([#7](https://github.com/vidiun/pakhshkit-js-kava/issues/7)) ([295d2d3](https://github.com/vidiun/pakhshkit-js-kava/commit/295d2d3))
* **FEC-8309:** playTimeSum value in View event is incorrect (not accumulative ([#11](https://github.com/vidiun/pakhshkit-js-kava/issues/11)) ([e58a9e5](https://github.com/vidiun/pakhshkit-js-kava/commit/e58a9e5))
* **FEC-8311:** Live with DVR - when seeking back (more then dvrThreshold) the playbackType is not "dvr" ([#12](https://github.com/vidiun/pakhshkit-js-kava/issues/12)) ([7e7be8b](https://github.com/vidiun/pakhshkit-js-kava/commit/7e7be8b))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/vidiun/pakhshkit-js-kava/compare/v0.2.0...v0.2.1) (2018-06-12)


### Bug Fixes

* **FEC-8304:** No VIEW events are sent ([#6](https://github.com/vidiun/pakhshkit-js-kava/issues/6)) ([9a87321](https://github.com/vidiun/pakhshkit-js-kava/commit/9a87321))



<a name="0.2.0"></a>

# [0.2.0](https://github.com/vidiun/pakhshkit-js-kava/compare/v0.1.0...v0.2.0) (2018-06-11)

### Features

- **FEC-8296:** add applicationVersion and playbackContext ([#3](https://github.com/vidiun/pakhshkit-js-kava/issues/3)) ([0252c59](https://github.com/vidiun/pakhshkit-js-kava/commit/0252c59))
- **FEC-8297:** reset plugin on critical error ([#4](https://github.com/vidiun/pakhshkit-js-kava/issues/4)) ([3a86cea](https://github.com/vidiun/pakhshkit-js-kava/commit/3a86cea))

<a name="0.1.0"></a>

# 0.1.0 (2018-06-05)

### Bug Fixes

- change model delegates to be on the instance ([abe2b2f](https://github.com/vidiun/pakhshkit-js-kava/commit/abe2b2f))
- set initial plugin version ([7bc3abb](https://github.com/vidiun/pakhshkit-js-kava/commit/7bc3abb))
- sources.type protection ([faf5875](https://github.com/vidiun/pakhshkit-js-kava/commit/faf5875))
- update sources type config ([3de325b](https://github.com/vidiun/pakhshkit-js-kava/commit/3de325b))

### Features

- kava plugin ([#1](https://github.com/vidiun/pakhshkit-js-kava/issues/1)) ([adf4609](https://github.com/vidiun/pakhshkit-js-kava/commit/adf4609))


test/src/kava.spec.js
import '../../src/index.js';
import {loadPlayer} from 'pakhshkit-js';
import * as TestUtils from 'pakhshkit-js/test/src/utils/test-utils';
import {OVPAnalyticsService, RequestBuilder} from 'pakhshkit-js-providers/dist/pakhshkit-analytics-service';
import {KavaEventModel} from '../../src/kava-event-model';

const targetId = 'player-placeholder_kava.spec';

describe('KavaPlugin', function() {
  let player;
  let kava;
  const config = {
    sources: {
      progressive: [
        {
          mimetype: 'video/mp4',
          url: 'http://www.html5videoplayer.net/videos/toystory.mp4'
        }
      ]
    },
    plugins: {
      kava: {}
    }
  };

  function createPlayerPlaceholder(targetId) {
    TestUtils.createElement('DIV', targetId);
    let el = document.getElementById(targetId);
    el.style.height = '360px';
    el.style.width = '640px';
  }

  function setupPlayer(config) {
    player = loadPlayer(config);
    const el = document.getElementById(targetId);
    el.appendChild(player.getView());
  }

  function getKavaPlugin() {
    return player._pluginManager.get('kava');
  }

  before(function() {
    createPlayerPlaceholder(targetId);
  });

  afterEach(function() {
    player.destroy();
    player = null;
    kava = null;
    TestUtils.removeVideoElementsFromTestPage();
  });

  after(function() {
    TestUtils.removeElement(targetId);
  });

  it('should reset kava session correctly', () => {
    setupPlayer(config);
    kava = getKavaPlugin();
    kava._resetSession();
    kava._model.getEventIndex().should.equal(1);
    kava._model.getBufferTimeSum().should.equal(0.0);
    kava._model.getPlayTimeSum().should.equal(0.0);
    kava._rateHandler.getAverage().should.equal(0);
  });

  it('should reset kava plugin correctly', () => {
    setupPlayer(config);
    kava = getKavaPlugin();
    kava.reset();
    kava._model.getEventIndex().should.equal(1);
    kava._model.getBufferTime().should.equal(0.0);
    kava._model.getBufferTimeSum().should.equal(0.0);
    kava._isFirstPlay.should.be.true;
    kava._isEnded.should.be.false;
    kava._isPaused.should.be.false;
    (kava._model.getSessionStartTime() === null).should.be.true;
    kava._timePercentEvent.PLAY_REACHED_25_PERCENT.should.be.false;
    kava._timePercentEvent.PLAY_REACHED_50_PERCENT.should.be.false;
    kava._timePercentEvent.PLAY_REACHED_75_PERCENT.should.be.false;
    kava._timePercentEvent.PLAY_REACHED_100_PERCENT.should.be.false;
  });

  describe('SendAnalytics', () => {
    let sandbox;
    const config = {
      sources: {
        progressive: [
          {
            bandwidth: 480256,
            mimetype: 'video/mp4',
            url: 'http://www.html5videoplayer.net/videos/toystory.mp4'
          }
        ]
      },
      plugins: {
        kava: {
          referrer: 'referrer',
          serviceUrl: '//analytics.vidiun.com/api_v3/index.php',
          viewEventCountdown: 10,
          resetSessionCountdown: 30,
          playerVersion: '0.18.1',
          playerName: 'vidiun-player-js',
          partnerId: '1091',
          entryId: '0_wifqaipd',
          entryType: 'Vod',
          sessionId: 'c15be273-0f1b-10a3-4fc9-d7a53eebee85:b66abd37-e2e2-a22e-86ac-7859592e754b',
          vs: 'Njk0ZmI4MzBiOTJiMGZhN2NmNTAwYWQyZGM2M2Y0YjkxMGRiZGI3MXwxMDkxOzEwOTE7MTUxNzkyMjgxMzswOzE1MTc4MzY0MTMuMTM4OzA7dmlldzoqLHdpZGdldDoxOzs='
        }
      },
      session: {
        id: 'c15be273-0f1b-10a3-4fc9-d7a53eebee85:b66abd37-e2e2-a22e-86ac-7859592e754b',
        partnerId: 1091,
        vs: 'Njk0ZmI4MzBiOTJiMGZhN2NmNTAwYWQyZGM2M2Y0YjkxMGRiZGI3MXwxMDkxOzEwOTE7MTUxNzkyMjgxMzswOzE1MTc4MzY0MTMuMTM4OzA7dmlldzoqLHdpZGdldDoxOzs='
      },
      id: '0_wifqaipd',
      name: 'MPEG Dash with MultiAudio New Transcoding',
      duration: 741,
      type: 'Vod',
      dvr: false
    };

    beforeEach(() => {
      sandbox = sinon.sandbox.create();
      sandbox.stub(RequestBuilder.prototype, 'doHttpRequest').callsFake(() => {
        return Promise.resolve();
      });
    });

    afterEach(() => {
      sandbox.restore();
    });

    function validateCommonParams(params, eventIndex) {
      params.eventType.should.equal(eventIndex);
      params.partnerId.should.equal(config.session.partnerId.toString());
      params.entryId.should.equal(config.id);
      params.sessionId.should.equal(config.session.id);
      params.eventIndex.should.equal(1);
      params.vs.should.equal(config.session.vs);
      params.referrer.should.equal(btoa(config.plugins.kava.referrer));
      params.deliveryType.should.equal('url');
      params.playbackType.should.equal('vod');
      (!!params.sessionStartTime).should.be.false;
      params.clientVer.should.equal(config.plugins.kava.playerVersion);
      params.clientTag.should.equal('html5:v' + config.plugins.kava.playerVersion);
      params.position.should.exist;
    }

    it('should send IMPRESSION event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.IMPRESSION.index) return;
        validateCommonParams(params, KavaEventModel.IMPRESSION.index);
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      player.play();
    });

    it('should send PLAY_REQUEST event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.PLAY_REQUEST.index) return;
        validateCommonParams(params, KavaEventModel.PLAY_REQUEST.index);
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      player.play();
    });

    it('should send PLAY event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.PLAY.index) return;
        validateCommonParams(params, KavaEventModel.PLAY.index);
        params.bufferTime.should.exist;
        params.bufferTimeSum.should.exist;
        params.actualBitrate.should.exist;
        params.joinTime.should.exist;
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      player.play();
    });

    it('should send RESUME event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.RESUME.index) return;
        validateCommonParams(params, KavaEventModel.RESUME.index);
        params.bufferTime.should.exist;
        params.bufferTimeSum.should.exist;
        params.actualBitrate.should.exist;
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      const onPlaying = () => {
        player.removeEventListener(player.Event.PLAYING, onPlaying);
        player.pause();
      };
      const onPause = () => {
        player.removeEventListener(player.Event.PLAYING, onPause);
        player.play();
      };
      player.addEventListener(player.Event.PLAYING, onPlaying);
      player.addEventListener(player.Event.PLAYING, onPause);
      player.play();
    });

    it('should send PAUSE event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.PAUSE.index) return;
        validateCommonParams(params, KavaEventModel.PAUSE.index);
        kava._timer._stopped.should.be.true;
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      const onPlaying = () => {
        player.removeEventListener(player.Event.PLAYING, onPlaying);
        player.pause();
      };
      player.addEventListener(player.Event.PLAYING, onPlaying);
      player.play();
    });

    it('should send REPLAY event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.REPLAY.index) return;
        validateCommonParams(params, KavaEventModel.REPLAY.index);
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      const onPlaying = () => {
        player.removeEventListener(player.Event.PLAYING, onPlaying);
        player.currentTime = player.duration;
      };
      const onEnded = () => {
        player.removeEventListener(player.Event.ENDED, onEnded);
        player.play();
      };
      player.addEventListener(player.Event.PLAYING, onPlaying);
      player.addEventListener(player.Event.ENDED, onEnded);
      player.play();
    });

    it('should send SEEK event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.SEEK.index) return;
        validateCommonParams(params, KavaEventModel.SEEK.index);
        params.targetPosition.should.exist;
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      const onPlaying = () => {
        player.removeEventListener(player.Event.PLAYING, onPlaying);
        player.currentTime = player.duration - 1;
      };
      player.addEventListener(player.Event.PLAYING, onPlaying);
      player.play();
    });

    it('should send PLAY_REACHED_25_PERCENT event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.PLAY_REACHED_25_PERCENT.index) return;
        validateCommonParams(params, KavaEventModel.PLAY_REACHED_25_PERCENT.index);
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      const onPlaying = () => {
        player.removeEventListener(player.Event.PLAYING, onPlaying);
        player.currentTime = player.duration / 4;
      };
      player.addEventListener(player.Event.PLAYING, onPlaying);
      player.play();
    });

    it('should send PLAY_REACHED_50_PERCENT event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.PLAY_REACHED_50_PERCENT.index) return;
        validateCommonParams(params, KavaEventModel.PLAY_REACHED_50_PERCENT.index);
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      const onPlaying = () => {
        player.removeEventListener(player.Event.PLAYING, onPlaying);
        player.currentTime = player.duration / 2;
      };
      player.addEventListener(player.Event.PLAYING, onPlaying);
      player.play();
    });

    it('should send PLAY_REACHED_75_PERCENT event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.PLAY_REACHED_75_PERCENT.index) return;
        validateCommonParams(params, KavaEventModel.PLAY_REACHED_75_PERCENT.index);
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      const onPlaying = () => {
        player.removeEventListener(player.Event.PLAYING, onPlaying);
        player.currentTime = player.duration - 1;
      };
      player.addEventListener(player.Event.PLAYING, onPlaying);
      player.play();
    });

    it('should send PLAY_REACHED_100_PERCENT event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.PLAY_REACHED_100_PERCENT.index) return;
        validateCommonParams(params, KavaEventModel.PLAY_REACHED_100_PERCENT.index);
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      const onPlaying = () => {
        player.removeEventListener(player.Event.PLAYING, onPlaying);
        player.currentTime = player.duration - 1;
      };
      player.addEventListener(player.Event.PLAYING, onPlaying);
      player.play();
    });

    it('should send SOURCE_SELECTED event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.SOURCE_SELECTED.index) return;
        validateCommonParams(params, KavaEventModel.SOURCE_SELECTED.index);
        params.actualBitrate.should.equal(480256);
        done();
        return new RequestBuilder();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      player.play();
      player.ready().then(() => {
        kava._sendAnalytics(KavaEventModel.SOURCE_SELECTED);
      });
    });

    it('should send FLAVOR_SWITCH event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.FLAVOR_SWITCH.index) return;
        validateCommonParams(params, KavaEventModel.FLAVOR_SWITCH.index);
        params.actualBitrate.should.equal(480256);
        done();
        return new RequestBuilder();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      player.play();
      player.ready().then(() => {
        kava._sendAnalytics(KavaEventModel.FLAVOR_SWITCH);
      });
    });

    it('should send AUDIO_SELECTED event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        if (params.eventType !== KavaEventModel.AUDIO_SELECTED.index) return;
        validateCommonParams(params, KavaEventModel.AUDIO_SELECTED.index);
        params.language.should.equal('heb');
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      kava._model.updateModel({language: 'heb'});
      kava._sendAnalytics(KavaEventModel.AUDIO_SELECTED);
    });

    it('should send CAPTIONS event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        validateCommonParams(params, KavaEventModel.CAPTIONS.index);
        params.caption.should.equal('eng');
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      kava._model.updateModel({caption: 'eng'});
      kava._sendAnalytics(KavaEventModel.CAPTIONS);
    });

    it('should send ERROR event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        validateCommonParams(params, KavaEventModel.ERROR.index);
        params.errorCode.should.equal(200);
        done();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      kava._model.updateModel({errorCode: 200});
      kava._sendAnalytics(KavaEventModel.ERROR);
    });

    it('should send VIEW event', done => {
      sandbox.stub(OVPAnalyticsService, 'trackEvent').callsFake((serviceUrl, params) => {
        validateCommonParams(params, KavaEventModel.VIEW.index);
        params.playTimeSum.should.exist;
        params.bufferTime.should.exist;
        params.bufferTimeSum.should.exist;
        params.actualBitrate.should.exist;
        params.averageBitrate.should.exist;
        done();
        return new RequestBuilder();
      });
      setupPlayer(config);
      kava = getKavaPlugin();
      player.play();
    });
  });

  describe('Server Response', () => {
    let sandbox;
    const config = {
      sources: {
        progressive: [
          {
            mimetype: 'video/mp4',
            url: 'http://www.html5videoplayer.net/videos/toystory.mp4'
          }
        ]
      },
      plugins: {
        kava: {
          serviceUrl: '//analytics.vidiun.com/api_v3/index.php',
          viewEventCountdown: 10,
          resetSessionCountdown: 30,
          playerVersion: '0.18.1',
          playerName: 'vidiun-player-js',
          partnerId: '1091',
          entryId: '0_wifqaipd',
          entryType: 'Vod',
          sessionId: 'c15be273-0f1b-10a3-4fc9-d7a53eebee85:b66abd37-e2e2-a22e-86ac-7859592e754b',
          vs: 'Njk0ZmI4MzBiOTJiMGZhN2NmNTAwYWQyZGM2M2Y0YjkxMGRiZGI3MXwxMDkxOzEwOTE7MTUxNzkyMjgxMzswOzE1MTc4MzY0MTMuMTM4OzA7dmlldzoqLHdpZGdldDoxOzs='
        }
      },
      session: {
        id: 'c15be273-0f1b-10a3-4fc9-d7a53eebee85:b66abd37-e2e2-a22e-86ac-7859592e754b',
        partnerId: 1091,
        vs: 'Njk0ZmI4MzBiOTJiMGZhN2NmNTAwYWQyZGM2M2Y0YjkxMGRiZGI3MXwxMDkxOzEwOTE7MTUxNzkyMjgxMzswOzE1MTc4MzY0MTMuMTM4OzA7dmlldzoqLHdpZGdldDoxOzs='
      },
      id: '0_wifqaipd',
      name: 'MPEG Dash with MultiAudio New Transcoding',
      duration: 741,
      type: 'Vod',
      dvr: false
    };

    beforeEach(() => {
      sandbox = sinon.sandbox.create();
    });

    afterEach(() => {
      sandbox.restore();
    });

    it('should handle server response of type number', done => {
      new Promise(resolve => {
        sandbox.stub(RequestBuilder.prototype, 'doHttpRequest').callsFake(() => {
          setTimeout(resolve, 1000);
          return Promise.resolve(12345);
        });
        setupPlayer(config);
        kava = getKavaPlugin();
        kava._sendAnalytics(KavaEventModel.VIEW);
      }).then(() => {
        kava._model.getSessionStartTime().should.equal(12345);
        done();
      });
    });

    it('should handle server response of type object', done => {
      new Promise(resolve => {
        sandbox.stub(RequestBuilder.prototype, 'doHttpRequest').callsFake(() => {
          setTimeout(resolve, 1000);
          return Promise.resolve({
            time: 12345,
            viewEventsEnabled: false
          });
        });
        setupPlayer(config);
        kava = getKavaPlugin();
        kava._sendAnalytics(KavaEventModel.VIEW);
      }).then(() => {
        kava._model.getSessionStartTime().should.equal(12345);
        kava._viewEventEnabled.should.equal(false);
        done();
      });
    });
  });
});


src/kava-timer.js
// @flow
import {FakeEvent, FakeEventTarget} from 'pakhshkit-js';

const SECOND: number = 1000;

/**
 * Mange a timer and dispatches related events.
 * @constructor
 * @param {Object} config - The timer config.
 */
class KavaTimer extends FakeEventTarget {
  _resetCounter: number;
  _eventCounter: number;
  _intervalId: ?number;
  _stopped: boolean;
  _config: Object;

  static Event = {
    TICK: 'tick',
    RESET: 'reset',
    REPORT: 'report'
  };

  constructor(config: Object) {
    super();
    this._config = config;
  }

  /**
   * Starts the timer interval.
   * @public
   * @returns {void}
   */
  start(): void {
    this._clearTimeout();
    this._stopped = false;
    this._resetCounter = 0;
    this._eventCounter = 0;
    this._intervalId = setInterval(() => this._monitor(), SECOND);
  }

  /**
   * Continues the timer to dispatch REPORT.
   * @public
   * @returns {void}
   */
  resume(): void {
    this._stopped = false;
    this._resetCounter = 0;
  }

  /**
   * Stops the timer from dispatch REPORT and starts the timer to dispatch RESET.
   * @public
   * @returns {void}
   */
  stop(): void {
    this._stopped = true;
  }

  /**
   * @public
   * @returns {boolean} - Whether the timer is stopped.
   */
  isStopped(): boolean {
    return this._stopped;
  }

  /**
   * Destroys the timer.
   * @public
   * @returns {void}
   */
  destroy(): void {
    this._clearTimeout();
  }

  _clearTimeout(): void {
    if (this._intervalId) {
      clearInterval(this._intervalId);
      this._intervalId = null;
    }
  }

  _monitor(): void {
    if (this._stopped) {
      if (this._resetCounter === this._config.resetCounter) {
        this.dispatchEvent(new FakeEvent(KavaTimer.Event.RESET));
        this._resetCounter = 0;
        this._eventCounter = 0;
      }
      this._resetCounter++;
    } else {
      this.dispatchEvent(new FakeEvent(KavaTimer.Event.TICK));
      if (this._eventCounter === this._config.eventCounter) {
        this.dispatchEvent(new FakeEvent(KavaTimer.Event.REPORT));
        this._eventCounter = 0;
      }
      this._eventCounter++;
    }
  }
}

export {KavaTimer};


src/kava-event-model.js
// @flow
import {KavaModel} from './kava-model';

export const KavaEventModel: {[event: string]: KavaEvent} = {
  VIEW: {
    type: 'VIEW',
    index: 99,
    getEventModel: (model: KavaModel) => ({
      playTimeSum: model.getPlayTimeSum(),
      bufferTime: model.getBufferTime(),
      bufferTimeSum: model.getBufferTimeSum(),
      actualBitrate: model.getActualBitrate(),
      averageBitrate: model.getAverageBitrate()
    })
  },
  IMPRESSION: {
    type: 'IMPRESSION',
    index: 1,
    getEventModel: () => ({})
  },
  PLAY_REQUEST: {
    type: 'PLAY_REQUEST',
    index: 2,
    getEventModel: () => ({})
  },
  PLAY: {
    type: 'PLAY',
    index: 3,
    getEventModel: (model: KavaModel) => ({
      bufferTime: model.getBufferTime(),
      bufferTimeSum: model.getBufferTimeSum(),
      actualBitrate: model.getActualBitrate(),
      joinTime: model.getJoinTime()
    })
  },
  RESUME: {
    type: 'RESUME',
    index: 4,
    getEventModel: (model: KavaModel) => ({
      bufferTime: model.getBufferTime(),
      bufferTimeSum: model.getBufferTimeSum(),
      actualBitrate: model.getActualBitrate()
    })
  },
  PAUSE: {
    type: 'PAUSE',
    index: 33,
    getEventModel: () => ({})
  },
  REPLAY: {
    type: 'REPLAY',
    index: 34,
    getEventModel: () => ({})
  },
  SEEK: {
    type: 'SEEK',
    index: 35,
    getEventModel: (model: KavaModel) => ({
      targetPosition: model.getTargetPosition()
    })
  },
  PLAY_REACHED_25_PERCENT: {
    type: 'PLAY_REACHED_25_PERCENT',
    index: 11,
    getEventModel: () => ({})
  },
  PLAY_REACHED_50_PERCENT: {
    type: 'PLAY_REACHED_50_PERCENT',
    index: 12,
    getEventModel: () => ({})
  },
  PLAY_REACHED_75_PERCENT: {
    type: 'PLAY_REACHED_75_PERCENT',
    index: 13,
    getEventModel: () => ({})
  },
  PLAY_REACHED_100_PERCENT: {
    type: 'PLAY_REACHED_100_PERCENT',
    index: 14,
    getEventModel: () => ({})
  },
  SOURCE_SELECTED: {
    type: 'SOURCE_SELECTED',
    index: 39,
    getEventModel: (model: KavaModel) => ({
      actualBitrate: model.getActualBitrate()
    })
  },
  AUDIO_SELECTED: {
    type: 'AUDIO_SELECTED',
    index: 42,
    getEventModel: (model: KavaModel) => ({
      language: model.getLanguage()
    })
  },
  FLAVOR_SWITCH: {
    type: 'FLAVOR_SWITCH',
    index: 43,
    getEventModel: (model: KavaModel) => ({
      actualBitrate: model.getActualBitrate()
    })
  },
  CAPTIONS: {
    type: 'CAPTIONS',
    index: 38,
    getEventModel: (model: KavaModel) => ({
      caption: model.getCaption()
    })
  },
  ERROR: {
    type: 'ERROR',
    index: 98,
    getEventModel: (model: KavaModel) => ({
      errorCode: model.getErrorCode()
    })
  }
};

/**
 * Gets the full event model for a certain event object including the common params.
 * @param {KavaEvent} eventObj - The event model.
 * @param {KavaModel} model - The plugin model store.
 * @returns {Object} - The full event model.
 */
export function getEventModel(eventObj: KavaEvent, model: KavaModel): Object {
  const commonModel: Object = {
    eventType: eventObj.index,
    partnerId: model.getPartnerId(),
    entryId: model.getEntryId(),
    sessionId: model.getSessionId(),
    eventIndex: model.getEventIndex(),
    referrer: model.getReferrer(),
    deliveryType: model.getDeliveryType(),
    playbackType: model.getPlaybackType(),
    clientVer: model.getClientVer(),
    clientTag: model.getClientTag(),
    position: model.getPosition()
  };
  if (model.getSessionStartTime()) {
    commonModel.sessionStartTime = model.getSessionStartTime();
  }
  if (model.getVS()) {
    commonModel.vs = model.getVS();
  }
  if (model.getUIConfId()) {
    commonModel.uiConfId = model.getUIConfId();
  }
  if (model.getCustomVar1()) {
    commonModel.customVar1 = model.getCustomVar1();
  }
  if (model.getCustomVar2()) {
    commonModel.customVar2 = model.getCustomVar2();
  }
  if (model.getCustomVar3()) {
    commonModel.customVar3 = model.getCustomVar3();
  }
  if (model.getPlaybackContext()) {
    commonModel.playbackContext = model.getPlaybackContext();
  }
  if (model.getApplicationVersion()) {
    commonModel.applicationVersion = model.getApplicationVersion();
  }
  const eventModel = eventObj.getEventModel(model);
  return Object.assign(eventModel, commonModel);
}
